applications/DG-Max/Algorithms/HarmonicSolver.cpp

```diff
@@ -212,7 +212,7 @@ HarmonicSolver<DIM>::Workspace::Workspace(
     // Initialize the basis functions and indices after creating the matrices,
     // so that the matrices are not yet assembled.
     discretization_->initializeBasisFunctions(*mesh_);
-    indexing_.reset(mesh_, Utilities::GlobalIndexing::BLOCKED_GLOBAL);
+    indexing_.reset(mesh_, Utilities::GlobalIndexing::BLOCKED_PROCESSOR);
```

Comment:
Here you set the indices for each processor instead of using the global indexing.

Reply:
There are solution coefficients (known as Degrees of Freedom) for each variable in the original problem and each mesh part (Element, Face, etc.). These options set in which order they should be enumerated, and BLOCKED_PROCESSOR, is not suitable to use with the requirements from regular PETSc matrices.

Reply:
The FEM (DG and CG) part I am very familiar with, my unfamiliarity is more with hpgem and PETSc. What do you mean by regular PETSc matrices? matseqaij? BLOCKED_PROCESSOR does indexing of the dofs per processor. BLOCKED_GLOBAL does the indexing globally, is that it?

Reply:
With the regular ones I mean the mpimaij like ones. But not the one where you create a block matrix out of individual smaller PETSc matrices.

So to reword the thing about indexing, MatSetValues uses the global row and column index to set or update values in the matrix. So we need to define a mapping from the basis function for element e, belonging to process p, and variable v (e.g. pressure, velocity) to the global row/column in the matrix. The difference between the three options is how to create a global ordering/index with these three subindices.

There are three options that I build at some point into hpgem, which order the basis functions lexicographically using the triples:
 - `SEQUENTIAL`: (p,e,v)
 - `BLOCKED_PROCESSOR`: (p,v,e)
 - `BLOCKED_GLOBAL`: (v,p,e)
So the `BLOCKED_GLOBAL` groups all basis functions of a single variable together, then orders them by process for that process orders them by element. As all the basis functions of a single variable are grouped together, the resulting global matrix would be a block matrix (per variable). However, this means that the indices of a single process are not continuous. This is not allowed by PETSc's MAT_MPIAIJ, which requires them to be sequential. So there `BLOCKED_PROCESSOR` comes in, this first orders them by process, so that all the rows/columns of a single process are consecutive, and within those rows creates the block structure of ordering by variable. Lastly, the `SEQUENTIAL` style does not create a larger block structure and for each process orders them by element.